### PR TITLE
Added Annotation Base Class

### PR DESCRIPTION
**Pull request by satya-pattnaik**
#### Reference Issues/PRs

With regards to #798 

#### What does this implement/fix? Explain your changes.

Creates a Base Class for Annotations.
#### What should a reviewer concentrate their feedback on?

If the Base Estimator contains the proper code and is in the proper directory.
##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [ ] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [ ] I've added unit tests and made sure they pass locally.

##### For new estimators
- [ ] I've added the estimator to the online documentation.
- [ ] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.

### CODE REVIEW

File: sktime/annotation/tests/test_all_annotators.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        Zt.iloc[1] = True
+        return Zt
+
+
+test_annotator = TestAnnotator()
+
+
+def test_output_is_series():
+
+    data = pd.Series(range(5))
```
Comment (mloning) [inline sktime/annotation/tests/test_all_annotators.py:26]:
```python
ANNOTATORS = all_estimators("annotator", return_names=False)

@pytest.mark.parametrize("Estimator", ANNOTATORS):
def my_test_function(Estimator):
  estimator = _construct_instance(Estimator)
  
  args = _make_args(estimator, "fit")
  estimator.fit(*args)
  # assert somthing

  args = _make_args(estimator, "transform")
  estimator.transform(*args)
  # assert something
```
* need to update `all_estimators(...)`
* need to update `_make_args(...)` to handle annotators as estimators

File: sktime/annotation/tests/test_all_annotators.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+# -*- coding: utf-8 -*-
+import pandas as pd
+import numpy as np
+from sktime.annotation.base._base import BaseAnnotator
+
+
+class TestAnnotator(BaseAnnotator):
```
Comment (mloning) [inline sktime/annotation/tests/test_all_annotators.py:7]: create mock annotator class in `sktime/annotation/base/_mock.py` to make sure all the unit testing functionality works
```python
class MockAnnotator(BaseAnnotator):
    # mock class, implement basic interface fit/predict
```

**Review by mloning — COMMENTED** [review]
Hi @satya-pattnaik as discussed in our meeting:

File: sktime/annotation/base/_mock.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+class MockAnnotator(BaseAnnotator):
+    def __init__(self):
+        pass
+
+    def fit(self, Z, X=None):
+        self._is_fitted = True
+        return self
+
+    def transform(self, Z, X=None):
```
Comment (mloning) [inline sktime/annotation/base/_mock.py:13]: Should be `predict` as in BaseAnnotator above

File: sktime/utils/_testing/estimator_checks.py [hunk 4]
```diff
@@ -556,6 +559,8 @@ def _make_fit_args(estimator, **kwargs):
         fh = 1
         X = None
         return y, X, fh
+    elif isinstance(estimator, BaseAnnotator):
```
Comment (mloning) [inline sktime/utils/_testing/estimator_checks.py:562]: Yes, this is good, we also need to modify `_make_predict_args` accordingly

```python
def _make_predict_args(estimator, **kwargs):    
    ...
    elif isinstance(estimator, BaseAnnotator):
        return make_annotation_problem(n_timepoints=10, **kwargs)
```

**Review by mloning — COMMENTED** [review]
Comments as discussed in our call. 

For the test on bad input arguments: 

```python
@pytest.mark.parametrize("arg", BAD_INPUT_ARGS)
def test_bad_input_raises_error(arg):
    msg = "expected error msg"
    with pytest.raises(ValueError, match=msg):
        check_X(arg)

```

**Comment by fkiraly** [comment]
@satya-pattnaik, @mloning, great work!

I would, however, suggest to think about the following:

* I would strongly suggest that the annotated time series `z` does *not* need to have the same indices!
* I don't think we need exogeneous data - can we not just assume that the series to segment is multivariate to start with? If the annotator insists to focus on one variable, then this can be passed as info to init.
* I would suggest to cover both supervised and unsupervised annotators, that should be easy with a "meta-base-class". Supervised ones accept a `Y` which are training annotations.
* I would suggest to call the object being annotated `X`, not `Z`.
* I would suggest to adopt the `fit/_fit` design from forecasters - this will make input checks etc easier. We'll probably move there too with TSC, so let's make this uniform!
* for segmentation, I would suggest the convention returns a series with an `pandas.IntervalIndex`; more generally, the docstrings should describe the convention of the input or return objects, if any - implement tests accordingly.
* I strongly suggest that we also think about the panel case - this is important, and covered ,methodologically by key annotation packages in the segmentation domain, like `hmmlearn`. The format I'd suggest is "list-of-X", and for supervised learners "list-of-Y" as a label. Such annotators should also accept the "single-sample" case.

**Comment by mloning** [comment]
Closed in favour of #1021

**Pull request closed without merging**